core/src/main/java/org/apache/iceberg/IncrementalDataScan.java

```diff
@@ -0,0 +1,118 @@
+    super(ops, table);
+    this.ops = ops;
+    this.table = table;
+    this.fromSnapshotId = fromSnapshotId;
+    this.toSnapshotId = toSnapshotId;
+  }
+
+  private IncrementalDataScan(TableOperations ops, Table table, Long snapshotId, Schema schema,
+                              Expression rowFilter, boolean caseSensitive, boolean colStats,
+                              Collection<String> selectedColumns, ImmutableMap<String, String> options,
+                              Long fromSnapshotId, Long toSnapshotId) {
+    super(ops, table, snapshotId, schema, rowFilter, caseSensitive, colStats, selectedColumns, options);
+    this.ops = ops;
+    this.table = table;
+    this.colStats = colStats;
+    this.fromSnapshotId = fromSnapshotId;
+    this.toSnapshotId = toSnapshotId;
+  }
+
+  @Override
+  public TableScan asOfTime(long timestampMillis) {
+    throw new UnsupportedOperationException("asOfTime() is not supported in IncrementalScan");
+  }
+
+  @Override
+  public TableScan useSnapshot(long scanSnapshotId) {
+    throw new UnsupportedOperationException("useSnapshot() is not supported in IncrementalScan");
+  }
+
+  @Override
+  public CloseableIterable<FileScanTask> planFiles() {
+    //TODO publish an incremental appends scan event
+    List<Snapshot> snapshots = snapshotsWithin(table, fromSnapshotId, toSnapshotId);
+    final Iterable<CloseableIterable<FileScanTask>> files = Iterables.transform(snapshots, this::planFiles);
+    return CloseableIterable.concat(files);
+  }
+
+  private CloseableIterable<FileScanTask> planFiles(Snapshot snapshot) {
+    Predicate<ManifestFile> matchingManifests = manifest -> manifest.snapshotId() == snapshot.snapshotId();
+
+    Predicate<ManifestEntry> matchingManifestEntries =
+        manifestEntry ->
+            manifestEntry.snapshotId() == snapshot.snapshotId() &&
+            manifestEntry.status() == ManifestEntry.Status.ADDED;
+
+    return planFiles(ops, snapshot, filter(), isCaseSensitive(), colStats, matchingManifests, matchingManifestEntries);
+  }
+
+  @Override
+  @SuppressWarnings("checkstyle:HiddenField")
+  protected TableScan newRefinedScan(
+          TableOperations ops, Table table, Long snapshotId, Schema schema, Expression rowFilter,
+          boolean caseSensitive, boolean colStats, Collection<String> selectedColumns,
+          ImmutableMap<String, String> options) {
+    return new IncrementalDataScan(
+            ops, table, snapshotId, schema, rowFilter, caseSensitive, colStats, selectedColumns, options,
+            fromSnapshotId, toSnapshotId);
+  }
+
+  private static List<Snapshot> snapshotsWithin(Table table, long fromSnapshotId, long toSnapshotId) {
+    List<Long> snapshotIds = SnapshotUtil.currentAncestors(table);
+    int toSnapshotIdIndex = snapshotIds.indexOf(toSnapshotId);
+    Preconditions.checkArgument(toSnapshotIdIndex != -1, "toSnapshotId: %s does not exist", toSnapshotId);
```

Comment:
If this is going to use `checkArgument`, then I think we should run the check earlier, when `appendsBetween` is called. In general, I think that `IllegalArgumentException` should be thrown in places where your stack trace includes the call that passed the bad argument in.

```java
  table.newScan()
      .appendsBetween(a, b) // <--- Should be thrown here
      .filter(...)
      .project("col1", "col2")
      .planSplits(); // <--- Currently thrown here
```

Also, the problem is that the given snapshot is not an ancestor of the current table state, not that it does not exist. We should fix the error message if we keep this behavior.

Should we require ancestors of the current table state? What if I want to use this to find the files changed on a WAP branch of a table? I might commit a snapshot without making it current, then use this to validate what was written. Then I'd actually scan `appendsBetween(currentSnapshot.snapshotId(), uncommitted.snapshotId())`. This should be find as long as `fromSnapshotId` is in the ancestors of `toSnapshotId`.